### ISSUE TITLE
v1.6.4 Initialization Error

### ISSUE THREAD

**Issue by santanu-biswas** [winstonjs/winston-mongodb#75]
Upgrading to v1.6.4 starts throwing the following error:

```
winston-mongodb, initialization error:  { MongoError: index not found with name [timestamp_1]
    at Function.MongoError.create (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/error.js:31:11)
    at /Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:483:72
    at authenticateStragglers (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:429:16)
    at Connection.messageHandler (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:463:5)
    at Socket.<anonymous> (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/connection.js:309:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:551:20)
  name: 'MongoError',
  message: 'index not found with name [timestamp_1]',
  nIndexesWas: 1,
  ok: 0,
  errmsg: 'index not found with name [timestamp_1]',
  code: 27 }
winston-mongodb, initialization error:  { MongoError: index not found with name [timestamp_1]
    at Function.MongoError.create (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/error.js:31:11)
    at /Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:483:72
    at authenticateStragglers (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:429:16)
    at Connection.messageHandler (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/pool.js:463:5)
    at Socket.<anonymous> (/Users/sbiswas/Documents/Code/SalesTracker/salestracker-web/node_modules/mongodb-core/lib/connection/connection.js:309:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:551:20)
  name: 'MongoError',
  message: 'index not found with name [timestamp_1]',
  nIndexesWas: 1,
  ok: 0,
  errmsg: 'index not found with name [timestamp_1]',
  code: 27 }
```

Checked relevant tables in MongoDB and the they do have the index named timestamp_1.

**Reverting back to v1.6.1 removes this error.** 
Looks like there is a bug in v1.6.4

**Comment by 39dotyt** [comment]
Will be fixed today. Thank you.

**Comment by 39dotyt** [comment]
@santanu-biswas checkout version 2.0.0

**Comment by santanu-biswas** [comment]
Bug resolved!
Thanks!

**Comment by perrin4869** [comment]
Hm... I'm having this problem still, running node 6.9.1 and mongodb version 3.0.6.

Here's the code used to add the mongodb transport:

```js
  if (config.logging.db) {
    logger.add(winston.transports.MongoDB, {
      db: db.getDatabasePromise(),
      collection: 'logs',
      level: config.logging.db,
    });
  }
```

The error I get:

```
Unhandled rejection MongoError: index not found with name [timestamp_1]
    at Function.MongoError.create (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/error.js:31:11)
    at /home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:483:72
    at authenticateStragglers (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:429:16)
    at Connection.messageHandler (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:463:5)
    at Socket.<anonymous> (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/connection.js:309:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
From previous event:
    at runCallback (timers.js:637:20)
    at tryOnImmediate (timers.js:610:5)
    at processImmediate [as _immediateCallback] (timers.js:582:5)
From previous event:
    at new exports.MongoDB (/home/perrin4869/projects/websites/serapi/node_modules/winston-mongodb/lib/winston-mongodb.js:151:13)
    at Logger.add (/home/perrin4869/projects/websites/serapi/node_modules/winston/lib/winston/logger.js:481:41)
    at module.exports (/home/perrin4869/projects/websites/serapi/logger.js:46:12)
    at Object.<anonymous> (/home/perrin4869/projects/websites/serapi/chat_server.js:23:35)
    at Module._compile (module.js:570:32)
    at loader (/home/perrin4869/projects/websites/serapi/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/home/perrin4869/projects/websites/serapi/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Function.Module._load (/home/perrin4869/projects/websites/serapi/node_modules/mock-require/index.js:14:22)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/perrin4869/projects/websites/serapi/bin/www:19:26)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

**Comment by 39dotyt** [comment]
@perrin4869 try upgrading to winston-mongodb@2

**Comment by 39dotyt** [comment]
@perrin4869 will backport fix to 1.6

**Comment by perrin4869** [comment]
@39dotyt I'm actually using `winston-mongodb@2`

**Comment by 39dotyt** [comment]
@perrin4869 what is `db.getDatabasePromise()`?

**Comment by perrin4869** [comment]
This function:

```js
function getDatabasePromise() {
  return new Promise((resolve) => {
    mongoose.connection.once('open', () => {
      resolve(mongoose.connection.db);
    });
  })
}
```

**Comment by perrin4869** [comment]
I have traced down the bug to the `createCollection` function. I'm using `mongoose@4.6.7`:

```js
      return col.dropIndex(ttlIndexName).catch(err=>{
        if (err.code !== 27) {
          throw err;
        }
      }).then(()=>col.createIndex({timestamp: -1}, indexOpts)
      ).then(()=>db);
```

In the snippet above from the function, `dropIndex` throws an error, but `err.code` is `undefined` and therefore the error is propagated. I have no clue why `err.code` is undefined though. It might be a problem with `mongoose`. But I did not have this problem before `winston-mongodb@2.0.0`.

**Comment by perrin4869** [comment]
OK, just confirmed the error was introduced in `v1.6.5`. `err.code` is undefined. I can also not find it in the documentation here: http://mongodb.github.io/node-mongodb-native/core/api/MongoError.html

**Comment by perrin4869** [comment]
This is the error thrown btw:

```
{ MongoError: index not found with name [timestamp_1]
    at Function.MongoError.create (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/error.js:31:11)
    at /home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:483:72
    at authenticateStragglers (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:429:16)
    at Connection.messageHandler (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/pool.js:463:5)
    at Socket.<anonymous> (/home/perrin4869/projects/websites/serapi/node_modules/mongodb-core/lib/connection/connection.js:339:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
  name: 'MongoError',
  message: 'index not found with name [timestamp_1]',
  nIndexesWas: 2,
  ok: 0,
  errmsg: 'index not found with name [timestamp_1]' }
```

**Comment by tibeoh** [comment]
I have the same issue...

**Comment by perrin4869** [comment]
@39dotyt I can try to put together a PR but I'm not sure what would be a good way to go about fixing this... any ideas?

**Comment by tibeoh** [comment]
Hi @perrin4869,
I agree with you. The issue is here: https://github.com/winstonjs/winston-mongodb/blob/master/lib/winston-mongodb.js#L117
The `error.code` doesn't exist in mongodb driver >= v2.0
```
// TODO: drop index only if params changed
return col.dropIndex(ttlIndexName).catch(err=>{
    if (err.code !== 27) {
       throw err;
    }
}).then(()=>col.createIndex({timestamp: -1}, indexOpts)
).then(()=>db);
```

I would like to fix issue but I don't understand why we need to drop index`timestamp`? It's the `createCollection` method, so the index doesn't exist, no? If you explain to me the goal of this drop, I can try to fix issue.

**Comment by Jeff-Lewis** [comment]
@tibeoh sounds like the issue. `createIndex()` [will not update an existing index by the same name](https://docs.mongodb.com/v3.2/reference/method/db.collection.createIndex/#behaviors). It must be dropped first. 

Since this behavior changed in mongodb 2.6 and also the node client might handle this differently, it's [safer to check](https://docs.mongodb.com/v3.2/reference/method/db.collection.getIndexes/#db.collection.getIndexes) `col.getIndexes()` first to see if it exists instead of relying on error and error codes.

**Comment by 39dotyt** [comment]
Fixed in 1.6.6 and 2.0.1.
